software/firmware/configuration.py

```diff
@@ -200,12 +201,80 @@ def load_config(cls, config_spec: ConfigSpec):
 
             config.update(saved_config)
-
-            return config
+            return ConfigSettings(config)
         else:
-            return {}
+            return ConfigSettings({})
 
     @staticmethod
     def delete_config(cls):
         """Deletes the config file, effectively resetting to defaults."""
         delete_file(ConfigFile.config_filename(cls))
+
+
+class ConfigSettings:
+    """Collects the configuration settings into an object with attributes instead of a dict with keys
+
+    Dict keys are converted to upper-case strings with underscores
+    """
+
+    def __init__(self, d):
+        """Constructor
+
+        @param d  The raw dict loaded from the configuration file
+        """
+        self.__dict__ = {}  # required for getattr & setattr
+
+        for k in d.keys():
+            cname = self.to_attr_name(k)
+            setattr(self, cname, d[k])
+
+    def to_attr_name(self, key):
+        """Converts a dict key string to its equivalent attribute name
+
+        @param key  The string to convert
+        @return     The same string, converted to upper-case, preserving underscores
+
+        @exception  ValueError if the key contains invalid characters; only letters, numbers, hyphens, and underscores
+                    are permitted. They key cannot be length 0, nor can it begin with a number
+        """
+        key = key.strip()
+        s = ""
+        for ch in key:
+            if ch.isalpha() or ch.isdigit() or ch == "_":
+                s += ch.upper()
+            else:
+                raise ValueError(
+                    f"Invalid attribute name: {key}. Keys cannot contain the character {ch}"
+                )
+
+        if len(s) == 0:
+            raise ValueError("Invalid attribute name: key cannot be empty")
+        elif s[0].isdigit():
+            raise ValueError("Invalid attribute name: key cannot start with a number")
+        return s
+
+    def __eq__(self, that):
+        """Allows comparing the config object directly to either another config object or a dict
+
+        @param that  The object we're comparing to, either a dict or another ConfigSettings object
+
+        @return True if the two objects are equivalent, otherwise False
+        """
+        if type(that) is dict:
+            try:
+                that = ConfigSettings(that)
+                return self == that
+            except ValueError:
+                return False
+        elif type(that) is ConfigSettings:
+            # Make sure every key self exists and is equal to the equivalent in that
+            for k in self.__dict__.keys():
+                if not k in that.__dict__ or self.__dict__[k] != that.__dict__[k]:
+                    return False
+
+            # Make sure every key in that exists and is equal to its equivalent in self
+            for k in that.__dict__.keys():
+                if not k in self.__dict__ or self.__dict__[k] != that.__dict__[k]:
```

Comment:
[optional] This ends up doing the comparison of every key twice. It would probably be more efficient to first do a comparison of the keys, and if and only if they match exactly do a comparison of the values.

That said, can we not just do a comparison of the underlying dicts here as well as is done above?

Also, it would be nice to see some tests for this kind of thing.

Reply:
This function literally only exists because of the CI tests, so efficiency wasn't really a concern. But you're right, comparing the two `__dict__` instances would be simpler.